File: cses/04_graph_algorithms/1197_cycle-finding/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    edges = []
    for _ in range(m):
        a = int(data[idx]); b = int(data[idx + 1]); c = int(data[idx + 2])
        idx += 3
        edges.append((a, b, c))

    # Khởi tạo dist = 0 cho MỌI đỉnh (tương đương thêm đỉnh nguồn ảo nối tới
    # tất cả các đỉnh với trọng số 0) để phát hiện chu trình âm ở bất kỳ đâu,
    # kể cả các chu trình không đến được từ đỉnh 1.
    dist = [0] * (n + 1)
    parent = [-1] * (n + 1)

    # Bellman-Ford: lặp đúng n vòng, mỗi vòng relax toàn bộ m cạnh.
    x = -1
    for _ in range(n):
        x = -1
        for a, b, c in edges:
            if dist[a] + c < dist[b]:
                dist[b] = dist[a] + c
                parent[b] = a
                x = b  # ghi lại đỉnh vừa được relax
        if x == -1:  # không còn relax được -> đã hội tụ -> không có chu trình âm
            break

    if x == -1:
        sys.stdout.write("NO\n")
        return

    # x được relax ở vòng thứ n => đi ngược theo parent n bước chắc chắn rơi
    # vào một đỉnh nằm trong chu trình âm.
    y = x
    for _ in range(n):
        y = parent[y]

    # Tái tạo chu trình theo chuỗi parent (hướng ngược), sau đó đảo lại.
    cycle = [y]
    cur = parent[y]
    while cur != y:
        cycle.append(cur)
        cur = parent[cur]
    cycle.append(y)
    cycle.reverse()

    sys.stdout.write("YES\n" + " ".join(map(str, cycle)) + "\n")


main()

File: cses/04_graph_algorithms/1197_cycle-finding/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

struct Edge {
    int source;
    int destination;
    long long weight;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    std::vector<Edge> edges(m);
    for (Edge& edge : edges) {
        std::cin >> edge.source >> edge.destination >> edge.weight;
    }

    // Khởi tạo dist = 0 cho MỌI đỉnh: tương đương thêm một đỉnh nguồn ảo nối
    // tới tất cả các đỉnh bằng cạnh trọng số 0, nhờ vậy phát hiện được chu
    // trình âm ở bất kỳ đâu, kể cả các thành phần không đến được từ đỉnh 1.
    std::vector<long long> distance(n + 1, 0);
    std::vector<int> parent(n + 1, -1);
    int changed_vertex = -1;

    // Bellman-Ford: lặp đúng n vòng, mỗi vòng relax toàn bộ m cạnh.
    for (int iteration = 0; iteration < n; ++iteration) {
        changed_vertex = -1;
        for (const Edge& edge : edges) {
            const long long candidate = distance[edge.source] + edge.weight;
            if (candidate < distance[edge.destination]) {
                distance[edge.destination] = candidate;
                parent[edge.destination] = edge.source;
                changed_vertex = edge.destination;  // ghi lại đỉnh vừa relax
            }
        }
        // Không còn relax được cạnh nào -> đã hội tụ -> không có chu trình âm.
        if (changed_vertex == -1) {
            break;
        }
    }

    // Còn relax được ở vòng thứ n nghĩa là chắc chắn có chu trình âm.
    if (changed_vertex == -1) {
        std::cout << "NO\n";
        return 0;
    }

    // Đỉnh vừa relax có thể chỉ đến được từ chu trình; đi ngược parent đúng n
    // bước để chắc chắn rơi vào một đỉnh nằm bên trong chu trình âm.
    int cycle_vertex = changed_vertex;
    for (int i = 0; i < n; ++i) {
        cycle_vertex = parent[cycle_vertex];
    }

    // Đi theo chuỗi parent từ đỉnh này cho tới khi quay lại nó để thu các đỉnh
    // của chu trình, rồi đảo lại vì parent biểu diễn hướng ngược.
    std::vector<int> cycle;
    int current = cycle_vertex;
    do {
        cycle.push_back(current);
        current = parent[current];
    } while (current != cycle_vertex);
    cycle.push_back(cycle_vertex);
    std::reverse(cycle.begin(), cycle.end());

    std::cout << "YES\n";
    for (int vertex : cycle) {
        std::cout << vertex << ' ';
    }
    std::cout << '\n';
    return 0;
}
